src/wasm/wasm-type.cpp

```diff
@@ -123,23 +123,18 @@ Type::Type(std::initializer_list<Type> types) { init(types); }
 
 Type::Type(const std::vector<Type>& types) { init(types); }
 
-size_t Type::size() const { return expand().size(); }
-
-const std::vector<Type>& Type::expand() const {
-  if (id <= _last_value_type) {
-    return basicTypes[id];
-  } else {
-    return *(std::vector<Type>*)id;
+size_t Type::size() const {
+  if (isMulti()) {
+    return (*(std::vector<Type>*)getID()).size();
   }
+  return size_t(getID() != Type::none);
```

Comment:
It's a little odd that we can't use `isSingle` here to make the code clearer. The more I think about it, the more it seems that `Type::unreachable` should actually be single, with the only types not being single being `Type::none` and tuples.

Reply:
Actually, I think the best thing to do here would be to change this to return 0 for unreachable as well as for none. This would be a functional change, but I think it would make more sense and I hope it wouldn't break anything. I think it's ok if an iterator over the unreachable type doesn't yield anything, so size() can remain consistent with the iterators.

Separately, it might make sense to put this logic in `Type::end()` and just implement `Type::size()` as `begin() - end()`.

Reply:
Moved the logic, but unfortunately found that making `unreachable` expand to `{}` leads to test errors - somewhat expected ofc. Also, one problem I see with this particular change is that we are essentially trading one odd check in `Type::end()` with special casing of `unreachable` all over the place, where currently code can just `expand()` respectively now use an iterator, not needing to worry about `unreachable`. Hmm...

Reply:
The test failures are unfortunate, but I think it's fine to leave this as a TODO for now. We can look into it in conjunction with #3062, which proposes getting rid of `unreachable` entirely.

Reply:
Just noticed https://github.com/WebAssembly/binaryen/issues/3062. Perhaps, with that being filed, it is fine to postpone a final decision here and just keep unreachable a special case until we know more?